File: server/src/errors/errorCodes.ts
/**
 * 统一错误码定义
 * 格式: 模块(2位) + 序号(3位)
 * 10xxx — Auth 模块
 * 20xxx — Location 模块
 * 30xxx — User 模块
 * 40xxx — Attendance 模块
 * 50xxx — Common/通用
 */
export const ErrorCodes = {
  // ===== Auth 模块 =====
  /** 手机号不能为空 */
  AUTH_PHONE_REQUIRED: { code: '10001', message: '手机号不能为空' },
  /** 密码不能为空 */
  AUTH_PASSWORD_REQUIRED: { code: '10002', message: '密码不能为空' },
  /** 手机号格式错误 */
  AUTH_PHONE_INVALID: { code: '10003', message: '手机号格式错误' },
  /** 密码长度不足 */
  AUTH_PASSWORD_TOO_SHORT: { code: '10004', message: '密码长度不能少于6位' },
  /** 账号或密码错误 */
  AUTH_LOGIN_FAILED: { code: '10005', message: '账号或密码错误' },
  /** Token无效或已过期 */
  AUTH_TOKEN_INVALID: { code: '10006', message: 'Token无效或已过期' },
  /** 未提供认证Token */
  AUTH_TOKEN_MISSING: { code: '10007', message: '未提供认证Token' },
  /** 用户不存在 */
  AUTH_USER_NOT_FOUND: { code: '10008', message: '用户不存在' },
  /** 无权限访问 */
  AUTH_FORBIDDEN: { code: '10009', message: '无权限访问' },

  // ===== Captcha / SMS 模块 =====
  /** 图形验证码错误 */
  CAPTCHA_CODE_ERROR: { code: '10015', message: '验证码错误' },
  /** 图形验证码已过期 */
  CAPTCHA_EXPIRED: { code: '10016', message: '验证码已过期，请重新获取' },
  /** 短信验证码错误 */
  SMS_CODE_ERROR: { code: '10017', message: '短信验证码错误' },
  /** 短信验证码已过期 */
  SMS_CODE_EXPIRED: { code: '10018', message: '短信验证码已过期，请重新获取' },
  /** 发送过于频繁 */
  SMS_TOO_FREQUENT: { code: '10019', message: '发送过于频繁，请60秒后再试' },
  /** 手机号格式不正确（需11位） */
  PHONE_FORMAT_INVALID: { code: '10020', message: '手机号格式不正确，请输入11位手机号' },

  // ===== Location 模块 =====
  /** 经度超出范围 */
  LOC_LNG_INVALID: { code: '20001', message: '经度超出有效范围(-180~180)' },
  /** 纬度超出范围 */
  LOC_LAT_INVALID: { code: '20002', message: '纬度超出有效范围(-85~85)' },
  /** 位置数据不能为空 */
  LOC_DATA_REQUIRED: { code: '20003', message: '位置数据不能为空' },
  /** 批量位置数据不能为空 */
  LOC_BATCH_EMPTY: { code: '20004', message: '批量上传数据不能为空' },
  /** 批量数据超过上限 */
  LOC_BATCH_TOO_LARGE: { code: '20005', message: '单次批量上传不能超过100条' },
  /** 日期参数无效 */
  LOC_DATE_INVALID: { code: '20006', message: '日期参数无效(格式: YYYY-MM-DD)' },

  // ===== Attendance 模块 =====
  /** 不在打卡范围内 */
  ATTEND_OUT_OF_RANGE: { code: '40001', message: '不在打卡范围内' },
  ATTEND_DUPLICATE: { code: '40005', message: '今日已打卡，请勿重复签到' },
  ATTEND_TIME_INVALID: { code: '40004', message: '不在打卡时间范围内' },

  // ===== Fence 模块 =====
  /** 围栏不存在 */
  FENCE_NOT_FOUND: { code: '30001', message: '围栏不存在' },
  /** 操作过于频繁 */
  FENCE_RATE_LIMITED: { code: '30002', message: '操作过于频繁，请稍后重试' },

  // ===== Org 模块 =====
  /** 用户已存在（手机号重复） */
  USER_EXISTS: { code: '30010', message: '该手机号已存在' },
  /** 数据库查询失败 */
  DB_ERROR: { code: '30011', message: '数据库查询失败' },
  /** 部门不存在 */
  DEPT_NOT_FOUND: { code: '30012', message: '部门不存在' },

  // ===== Upload 模块 =====
  /** 请求参数无效 */
  BAD_REQUEST: { code: '40010', message: '请求参数无效' },
  /** 文件类型不支持 */
  INVALID_FILE: { code: '40011', message: '不支持的文件类型' },

  // ===== Visit 模块 =====
  /** 拜访记录不存在 */
  VISIT_NOT_FOUND: { code: '30020', message: '拜访记录不存在' },
  /** 拜访已签到或已完成 */
  VISIT_ALREADY_CHECKIN: { code: '30021', message: '已签到或已完成' },
  /** 拜访已完成或已取消 */
  VISIT_COMPLETED: { code: '30022', message: '拜访已完成或已取消' },
  /** 未在拜访中 */
  VISIT_NOT_IN_PROGRESS: { code: '30023', message: '未在拜访中' },

  // ===== Message 模块 =====
  /** 消息不存在 */
  MESSAGE_NOT_FOUND: { code: '30030', message: '消息不存在' },

  // ===== Attendance Rules 模块 =====
  /** 考勤规则不存在 */
  ATTEND_RULE_NOT_FOUND: { code: '40006', message: '规则不存在' },

  // ===== Auth rate limit 模块 =====
  /** 操作过于频繁 */
  AUTH_RATE_LIMITED: { code: '10030', message: '操作过于频繁，请稍后重试' },
  /** 注册操作过于频繁 */
  REGISTER_RATE_LIMITED: { code: '10031', message: '注册操作过于频繁，请稍后再试' },

  // ===== Common 模块 =====
  /** 服务器内部错误 */
  INTERNAL_ERROR: { code: '50000', message: '服务器内部错误' },
  /** 请求参数无效 */
  PARAM_INVALID: { code: '50001', message: '请求参数无效' },
} as const;

export type ErrorCode = keyof typeof ErrorCodes;

/** 根据错误码获取状态码 */
export function getHttpStatus(errorCode: ErrorCode): number {
  const module = errorCode.split('_')[0];
  switch (module) {
    case 'AUTH':
      if (errorCode === 'AUTH_FORBIDDEN') return 403;
      if (errorCode === 'AUTH_TOKEN_MISSING' || errorCode === 'AUTH_TOKEN_INVALID' || errorCode === 'AUTH_LOGIN_FAILED') return 401;
      return 400;
    case 'CAPTCHA':
    case 'SMS':
    case 'PHONE':
    case 'LOC':
    case 'ATTEND':
    case 'FENCE':
    case 'VISIT':
      return 400;
    default:
      return 500;
  }
}

File: app/lib/services/error_codes.dart
/// 后端业务错误码映射表
///
/// 后端约定：成功响应 code=200，业务异常 code 范围：
///   Auth 10001~10009, 验证码 10015~10020, 位置 20001~20006,
///   Attendance 40001, Common 50000~50001
/// ApiService 拦截器自动拦截非 200 的 code 并转为 ApiException，
/// 业务层无需重复处理。
class ErrorCode {
  // ──────────────────────────────────────────────
  // Auth 模块（10001-10009）
  // ──────────────────────────────────────────────

  /// 10001 - 手机号不能为空
  static const int phoneRequired = 10001;

  /// 10002 - 密码不能为空
  static const int passwordRequired = 10002;

  /// 10003 - 手机号格式错误
  static const int phoneInvalid = 10003;

  /// 10004 - 密码长度不能少于6位
  static const int passwordTooShort = 10004;

  /// 10005 - 账号或密码错误
  static const int loginFailed = 10005;

  /// 10006 - Token无效或已过期
  static const int tokenInvalid = 10006;

  /// 10007 - 未提供认证Token
  static const int tokenMissing = 10007;

  /// 10008 - 用户不存在
  static const int userNotFound = 10008;

  /// 10009 - 无权限访问
  static const int forbidden = 10009;

  // ──────────────────────────────────────────────
  // Captcha / SMS 模块（10015-10020）
  // ──────────────────────────────────────────────

  /// 10015 - 验证码错误
  static const int captchaCodeError = 10015;

  /// 10016 - 验证码已过期
  static const int captchaExpired = 10016;

  /// 10017 - 短信验证码错误
  static const int smsCodeError = 10017;

  /// 10018 - 短信验证码已过期
  static const int smsCodeExpired = 10018;

  /// 10019 - 发送过于频繁
  static const int smsTooFrequent = 10019;

  /// 10020 - 手机号格式不正确
  static const int phoneFormatInvalid = 10020;

  // ──────────────────────────────────────────────
  // Location 模块（20001-20006）
  // ──────────────────────────────────────────────

  /// 20001 - 经度超出有效范围
  static const int lngInvalid = 20001;

  /// 20002 - 纬度超出有效范围
  static const int latInvalid = 20002;

  /// 20003 - 位置数据不能为空
  static const int locDataRequired = 20003;

  /// 20004 - 批量位置数据不能为空
  static const int locBatchEmpty = 20004;

  /// 20005 - 批量数据超过上限
  static const int locBatchTooLarge = 20005;

  /// 20006 - 日期参数无效
  static const int locDateInvalid = 20006;

  // ──────────────────────────────────────────────
  // Attendance 模块
  // ──────────────────────────────────────────────

  /// 40001 - 不在打卡范围内
  static const int attendOutOfRange = 40001;

  /// 40004 - 当前时间不在打卡有效时段内
  static const int attendTimeInvalid = 40004;

  /// 40005 - 今日已打卡，请勿重复签到
  static const int attendDuplicate = 40005;

  // ──────────────────────────────────────────────
  // Common 模块（50000-50001）
  // ──────────────────────────────────────────────

  /// 50000 - 服务器内部错误
  static const int internalError = 50000;

  /// 50001 - 请求参数无效
  static const int paramInvalid = 50001;

  // ──────────────────────────────────────────────
  // 错误码 → 中文提示 映射表
  // ──────────────────────────────────────────────

  static const Map<int, String> _messages = {
    // ── Auth ──
    10001: '手机号不能为空',
    10002: '密码不能为空',
    10003: '手机号格式错误',
    10004: '密码长度不能少于6位',
    10005: '账号或密码错误',
    10006: 'Token无效或已过期',
    10007: '未提供认证Token',
    10008: '用户不存在',
    10009: '无权限访问',
    // ── 预留 10010~10014（前端通用兜底）──
    10010: '系统繁忙，请稍后重试',
    10011: '请勿重复操作',
    10012: '数据已存在，无法重复添加',
    10013: '签到位置超出打卡范围',
    10014: '当前时间不在打卡有效时段内',
    // ── Captcha / SMS ──
    10015: '验证码错误',
    10016: '验证码已过期，请重新获取',
    10017: '短信验证码错误',
    10018: '短信验证码已过期，请重新获取',
    10019: '发送过于频繁，请60秒后再试',
    10020: '手机号格式不正确，请输入11位手机号',
    // ── Location ──
    20001: '经度超出有效范围(-180~180)',
    20002: '纬度超出有效范围(-85~85)',
    20003: '位置数据不能为空',
    20004: '批量上传数据不能为空',
    20005: '单次批量上传不能超过100条',
    20006: '日期参数无效(格式: YYYY-MM-DD)',
    // ── Attendance ──
    40001: '不在打卡范围内，请靠近打卡点',
    40004: '当前时间不在打卡有效时段内',
    40005: '今日已打卡，请勿重复签到',
    // ── Common ──
    50000: '服务器内部错误',
    50001: '请求参数无效',
  };

  /// 根据错误码获取用户友好提示
  static String message(int code) {
    return _messages[code] ?? '未知错误($code)';
  }

  /// 判断是否为业务层的错误码
  /// Auth(10001-10009), 预留(10010-10014), Captcha/SMS(10015-10020),
  /// Location(20001-20006), Attendance(40001), Common(50000-50001)
  static bool isBusinessError(int code) {
    return (code >= 10001 && code <= 10020) ||
        (code >= 20001 && code <= 20006) ||
        code == 40001 || code == 40004 || code == 40005 ||
        (code >= 50000 && code <= 50001);
  }

  /// 判断是否与登录/认证相关
  static bool isAuthError(int code) {
    return switch (code) {
      10002 || 10003 || 10004 || 10005 || 10007 => true,
      _ => false,
    };
  }

  /// 判断是否与签到业务相关
  static bool isAttendanceError(int code) {
    return switch (code) {
      10011 || 10012 || 10013 || 10014 || 40001 || 40004 || 40005 => true,
      _ => false,
    };
  }
}

/// API 业务异常，携带后端返回的错误码
class ApiException implements Exception {
  final int code;
  final String message;
  final Map<String, dynamic>? rawData;

  const ApiException({
    required this.code,
    this.message = '',
    this.rawData,
  }) : assert(code != 0, 'error code must not be 0');

  String get friendlyMessage =>
      message.isNotEmpty ? message : ErrorCode.message(code);

  @override
  String toString() => 'ApiException($code): $friendlyMessage';
}
